File: io.openems.backend.common/src/io/openems/backend/common/jsonrpc/request/SubmitSetupProtocolRequest.java
package io.openems.backend.common.jsonrpc.request;

import com.google.gson.JsonObject;

import io.openems.common.exceptions.OpenemsError.OpenemsNamedException;
import io.openems.common.jsonrpc.base.JsonrpcRequest;
import io.openems.common.utils.JsonUtils;

/**
 * Submits the Setup Protocol.
 *
 * <pre>
 * {
 *   "jsonrpc": "2.0",
 *   "id": UUID,
 *   "method": "submitSetupProtocol",
 *   "params": {
 *     "protocol": {
 *       "edge": {
 *         "id": string
 *       },
 *       "customer": {
 *         "firstname": string,
 *         "lastname": string,
 *         "email": string,
 *         "phone": string,
 *         "address": {
 *           "street": string,
 *           "city": string,
 *           "zip": string,
 *           "country": string
 *         },
 *         "company": {
 *          "name": string
 *         }
 *       },
 *       "location": {
 *         "firstname": string,
 *         "lastname": string,
 *         "email": string,
 *         "phone": string,
 *         "address": {
 *           "street": string,
 *           "city": string,
 *           "zip": string,
 *           "country": string
 *         },
 *         "company": {
 *          "name": string
 *         }
 *       },
 *       "lots": [{
 *         "category": string,
 *         "name": string,
 *         "serialNumber": string
 *       }],
 *       "items": [{
 *         "category": string,
 *         "name": string,
 *         "value": string
 *       }],
 *       "oem": string
 *     }
 *   }
 * }
 * </pre>
 */
public class SubmitSetupProtocolRequest extends JsonrpcRequest {

	public static final String METHOD = "submitSetupProtocol";

	/**
	 * Create {@link SubmitSetupProtocolRequest} from a template
	 * {@link JsonrpcRequest}.
	 *
	 * @param request the template {@link JsonrpcRequest}
	 * @return Created {@link SubmitSetupProtocolRequest}
	 * @throws OpenemsNamedException on parse error
	 */
	public static SubmitSetupProtocolRequest from(JsonrpcRequest request) throws OpenemsNamedException {
		var params = request.getParams();
		return new SubmitSetupProtocolRequest(request, JsonUtils.getAsJsonObject(params, "protocol"));
	}

	private final JsonObject jsonObject;

	private SubmitSetupProtocolRequest(JsonrpcRequest request, JsonObject jsonObject) {
		super(request, SubmitSetupProtocolRequest.METHOD);
		this.jsonObject = jsonObject;
	}

	@Override
	public JsonObject getParams() {
		return this.jsonObject;
	}

	/**
	 * Gets the Setup Protocol information as {@link JsonObject}.
	 *
	 * @return the {@link JsonObject}
	 */
	public JsonObject getJsonObject() {
		return this.jsonObject;
	}

}

File: ui/src/app/shared/jsonrpc/request/submitSetupProtocolRequest.ts
import { TranslateService } from "@ngx-translate/core";
import { Category } from "src/app/edge/installation/shared/category";
import { Utils } from "../../shared";
import { JsonrpcRequest } from "../base";

export type SetupProtocol = {
    edge: {
        id: string
    },
    installer: {
        firstname: string,
        lastname: string
    },
    customer: {
        firstname: string,
        lastname: string,
        email: string,
        phone: string,
        address: {
            street: string,
            city: string,
            zip: string,
            country: string
        },
        company?: {
            name: string
        }
    },
    location?: {
        firstname: string,
        lastname: string,
        email: string,
        phone: string,
        address: {
            street: string,
            city: string,
            zip: string,
            country: string
        },
        company?: {
            name: string

        }
    },
    lots?: {
        category: string,
        name: string,
        serialNumber: string
    }[],
    items?: {
        category: Category,
        name: string,
        value: string
    }[],
    oem: string
};

/**
 * <pre>
 * {
 *   "jsonrpc": "2.0",
 *   "id": UUID,
 *   "method": "submitSetupProtocol",
 *   "params": {
 *      protocol: {
 *          edge: {
 *              id: string
 *          },
 *          customer: {
 *              firstname: string,
 *              lastname: string,
 *              email: string,
 *              phone: string,
 *              address: {
 *                  street: string,
 *                  city: string,
 *                  zip: string,
 *                  country: string
 *              },
 *              company?: {
 *                  name: string
 *              }
 *          },
 *          location?: {
 *              firstname: string,
 *              lastname: string,
 *              email: string,
 *              phone: string,
 *              address: {
 *                  street: string,
 *                  city: string,
 *                  zip: string,
 *                  country: string
 *              },
 *              company?: {
 *                  name: string
 *              }
 *          },
 *          lots?: {
 *              category: string,
 *              name: string,
 *              serialNumber: string
 *          }[],
 *          items?: {
 *              category: Category,
 *              name: string,
 *              value: string
 *          }[],
 *          oem: string
 *      }
 *  }
 * </pre>
 */
export class SubmitSetupProtocolRequest extends JsonrpcRequest {

    static METHOD: string = "submitSetupProtocol";

    public static translateFrom(protocol: SetupProtocol, translate: TranslateService): SubmitSetupProtocolRequest {
        // protocol.items are type category in the protocol recieved and need to be translated before the request being sent.
        var items: {
            category: string,
            name: string,
            value: string
        }[] = protocol.items.map((element) => {
            return {
                category: Category.toTranslatedString(element.category, translate),
                name: element.name,
                value: element.value
            }
        })

        // 'Deep copy' to copy the object values from protocol recieved.
        // To avoid type issues from category to string.
        var protocolTranslated = Utils.deepCopy(protocol);
        protocolTranslated.items = items;

        return new SubmitSetupProtocolRequest({ protocol: protocolTranslated });
    }

    private constructor(
        public readonly params: {
            protocol: any
        }
    ) {
        super(SubmitSetupProtocolRequest.METHOD, params);
    }
}